igvc_gazebo/nodes/sim_color_detector/main.cpp

```diff
@@ -0,0 +1,219 @@
+
+  for(int rowCount=0; rowCount<frame.rows; ++rowCount) {
+    for(int columnCount=0; columnCount<frame.cols; ++columnCount)  {
+
+      cv::Vec3b color = frame.at<cv::Vec3b>(cv::Point(columnCount,rowCount));
+
+      if (color_check(color, PURPLE)) {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = white_color;
+      } else {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = black_color;
+      }   
+    }
+  }
+
+  sensor_msgs::Image outmsg;
+  outmsg.header = msg->header;
+  cv_ptr->image = output;
+  cv_ptr->encoding = "mono8";
+  cv_ptr->toImageMsg(outmsg);
+  outmsg_lines = outmsg;
+}
+
+/*
+ * Responsible for detecting the barrels (white and orange components) and
+ * illuminating them as white (and setting all other pixels to black).
+ */
+void handle_image_barrels(const sensor_msgs::ImageConstPtr& msg) {
+
+  cv_bridge::CvImagePtr cv_ptr;
+  cv::Mat frame;
+
+  try {
+    cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
+  } catch (cv_bridge::Exception& e) {
+    ROS_ERROR("CV-Bridge error: %s", e.what());
+    return;
+  }
+
+  frame = cv_ptr->image; // Input image
+  cv::Mat output(frame.rows, frame.cols, CV_8UC1, cv::Scalar::all(0)); // Ouput image (B&W)
+
+  const int white_color = 255;
+  const int black_color = 0;
+
+  for(int rowCount=0; rowCount<frame.rows; ++rowCount) {
+    for(int columnCount=0; columnCount<frame.cols; ++columnCount)  {
+
+      cv::Vec3b color = frame.at<cv::Vec3b>(cv::Point(columnCount,rowCount));
+
+      if (color_check(color, BLACK)) {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = white_color;
+      } else if (color_check(color, ORANGE)) {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = white_color;
+      } else if (color_check(color, WHITE)) {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = white_color;
+      } else {
+        output.at<uchar>(cv::Point(columnCount, rowCount)) = black_color;
+      }   
+    }
+  }
+
+  sensor_msgs::Image outmsg;
+  outmsg.header = msg->header;
+  cv_ptr->image = output;
+  cv_ptr->encoding = "mono8";
+  cv_ptr->toImageMsg(outmsg);
+  outmsg_barrels = outmsg;
+}
+
+/*
+ * Takes the most recent messages from the handle_image_lines and
+ * handle_image_barrels functions and publishes them once triggered
+ * by a a ros::Timer
+ */
+void publish(const ros::TimerEvent&)
+{
+  img_pub_barrels.publish(outmsg_barrels);
+  img_pub_lines.publish(outmsg_lines);
+}
+
+/*
+ * Responsible for setting up the publishers and triggering the functions to
+ * update the detected black-and-white images
+ */
+int main(int argc, char** argv) {
+
+  ros::init(argc, argv, "sim_color_detector");
+  
+  // LINES
+
+  ros::NodeHandle nh_lines;
+  ros::NodeHandle pNh_lines("~");
+
+  std::string topic_name_lines;
+  
+  igvc::param(pNh_lines, "image_topic", topic_name_lines, std::string("/center_cam/image_raw"));
+
+  img_pub_lines = nh_lines.advertise<sensor_msgs::Image>("/usb_cam_center/detected_lines", 1);
```

Comment:
Change the toppic advertised to something based on `topic_name_lines`, perhaps `topic_name_lines + "/detected_lines"` or something.